[Core] Reformat structure string operators

aaronfranke opened the pull request:
Implements and closes https://github.com/godotengine/godot-proposals/issues/1819

Explanation in bullets below each set of outputs. Test code:

```gdscript
func _ready():
	print("Vector2")
	print(Vector2())
	print("Vector2i")
	print(Vector2i())
	print("Vector3")
	print(Vector3())
	print("Vector3i")
	print(Vector3i())
	print("Color")
	print(Color())
	print("Quaternion")
	print(Quaternion())
	print("Plane")
	print(Plane())
	print("AABB")
	print(AABB())
	print("Rect2")
	print(Rect2())
	print("Rect2i")
	print(Rect2i())
	var t = Transform2D()
	t.x = Vector2(1, 5)
	print("Transform2D")
	print(t)
	var b = Basis()
	b.x = Vector3(1, 5, 0)
	print("Basis")
	print(b)
	print("Transform3D")
	print(Transform(b))
```

Output with current master:

```
Vector2
(0, 0)
Vector2i
(0, 0)
Vector3
(0, 0, 0)
Vector3i
(0, 0, 0)
Color
0,0,0,1
Quaternion
(0, 0, 0, 1)
Plane
0, 0, 0, 0
AABB
0, 0, 0 - 0, 0, 0
Rect2
(0, 0, 0, 0)
Rect2i
(0, 0, 0, 0)
Transform2D
((1, 5), (0, 1), (0, 0))
Basis
((1, 0, 0), (5, 1, 0), (0, 0, 1))
Transform3D
1, 5, 0, 0, 1, 0, 0, 0, 1 - 0, 0, 0
```

The current output is a bit confusing and inconsistent:

* Color is lacking spaces after the commas.

* AABB and Transform both contain a dash, which can be mistaken for a minus sign.

* Some things have parenthesis, some don't, and Basis/Transform2D have nested parenthesis.

* Transform2D prints out column-major, while Basis prints out row-major, and Transform is row major except the origin is last.

* The current internal code is a bit strange, with some of the string logic being inside of Variant and some being in the class.

Output with this PR:

```
Vector2
(0, 0)
Vector2i
(0, 0)
Vector3
(0, 0, 0)
Vector3i
(0, 0, 0)
Color
(0, 0, 0, 1)
Quaternion
(0, 0, 0, 1)
Plane
[N: (0, 0, 0), D: 0]
AABB
[P: (0, 0, 0), S: (0, 0, 0)]
Rect2
[P: (0, 0), S: (0, 0)]
Rect2i
[P: (0, 0), S: (0, 0)]
Transform2D
[X: (1, 5), Y: (0, 1), O: (0, 0)]
Basis
[X: (1, 5, 0), Y: (0, 1, 0), Z: (0, 0, 1)]
Transform3D
[X: (1, 5, 0), Y: (0, 1, 0), Z: (0, 0, 1), O: (0, 0, 0)]
```

* I've moved all the code to the classes and out of Variant.

* No dashes anywhere, parenthesis used for grouping with commas for separation.

* Transforms and Basis are now printed consistently and with explicit labels to avoid any confusion.

* Since they store `real_t`, I decided to use `String::num_real`, but normally this adds a trailing `.0` to whole numbers, so I added a new bool to include it or not called `p_trailing`, which defaults to true so its current uses are the same.

* Color is an exception to the above, it uses `String::num` with a fixed amount of decimals (4 seems like a good value, but maybe 5, 3 or 2 would be better).

aaronfranke commented:
I updated this PR with C# changes that match the behavior above. Note that since building with Mono doesn't work right now, I tested these changes by creating an empty .NET Core project and then copy-pasting these files into it. And yes, it works fine.

This should be good to merge any time, the sooner the better.

akien-mga commented on core/ustring.cpp:
```diff
@@ -1325,8 +1325,10 @@ String String::num_real(double p_num) {
 		}
 		sd = '.' + decimal;
-	} else {
+	} else if (p_trailing) {
```
Do we really want to make this optional though? IMO it would be just as well to enforce it, so that it's always clear that it's a real and not an integer.

aaronfranke replied:
Either is fine to me. I made it this way because it looks cleaner, but maybe it's better to always show it to make it clear that it's a real.

P.S. In C#, the default behavior of converting numbers to string does not include the trailing decimals, so if we want them I'd add a format specifier (should be something like `"0.0"`).

jonbonazza commented:
@akien-mga @aaronfranke i put this in the accompanying proposal already, but this is important so im leaving it here too. This can't be merged as-is

Printing mtiple lines is npt an option. Often times, you want to coalesce the string representation into another string. Most importantly though, it makes it unusable in log lines (for instance in server logs that are consumed by a log consumer such as fluentd)

aaronfranke committed:
Reformat structure string operators

The order of numbers is not changed except for Transform2D. All logic is done inside of their structures (and not in Variant).

For the number of decimals printed, they now use String::num_real which works best with real_t, except for Color which is fixed at 4 decimals (this is a reliable number of float digits when converting from 16-bpc so it seems like a good choice)

akien-mga commented on core/math/transform_2d.cpp:
```diff
@@ -277,5 +277,7 @@ Transform2D Transform2D::interpolate_with(const Transform2D &p_transform, real_t
 Transform2D::operator String() const {
-	return String(String() + elements[0] + ", " + elements[1] + ", " + elements[2]);
+	return "[X: " + elements[0].operator String() +
+		   ", Y: " + elements[1].operator String() +
+		   ", O: " + elements[2].operator String() + "]";
```
You seem to have space indentation/alignment here (and in other classes). It's annoying that `clang-format` allows it even though we use tabs only.

Does it work with a double-tab continuation indent instead?

aaronfranke replied:
@akien-mga It doesn't. `clang-format` automatically changes it back to the current state.

akien-mga commented:
Thanks!
